common/multiplexer.go

```diff
@@ -0,0 +1,63 @@
+package common
+
+import (
+	"context"
+	"sync"
+
+	"go.opentelemetry.io/otel/trace"
+)
+
+// Multiplexer provides a way to deduplicate concurrent operations with the same key.
+// It ensures that only one operation is executed while others wait for the result.
+// Type parameter T represents the result type of the operation.
+type Multiplexer[T any] struct {
+	key    string
+	result T
+	err    error
+	done   chan struct{}
+	mu     sync.RWMutex
+	once   sync.Once
+}
+
+// NewMultiplexer creates a new multiplexer instance with the given key.
+func NewMultiplexer[T any](key string) *Multiplexer[T] {
+	return &Multiplexer[T]{
+		key:  key,
+		done: make(chan struct{}),
+	}
+}
+
+// Close signals that the operation is complete and provides the result and error.
+// It ensures that Close is only called once, even if called from multiple goroutines.
+func (m *Multiplexer[T]) Close(ctx context.Context, result T, err error) {
+	_, span := StartDetailSpan(ctx, "Multiplexer.Close",
+		trace.WithAttributes(),
+	)
+	defer span.End()
+
+	m.once.Do(func() {
+		m.mu.Lock()
+		defer m.mu.Unlock()
+		m.result = result
+		m.err = err
+		close(m.done)
+	})
+}
+
+// Done returns a channel that is closed when the operation completes.
+func (m *Multiplexer[T]) Done() <-chan struct{} {
+	return m.done
+}
+
+// Result returns the result and error from the completed operation.
+// It should only be called after Done() channel is closed.
+func (m *Multiplexer[T]) Result() (T, error) {
+	m.mu.RLock()
+	defer m.mu.RUnlock()
+	return m.result, m.err
+}
+
+// Key returns the key associated with this multiplexer.
+func (m *Multiplexer[T]) Key() string {
+	return m.key
```

Comment:
Any reason for not having `key` field just to be public (vs a getter func)?

Reply:
hmm I thought the key here was meant to be immutable (only being set in the constructor?)

Reply:
Yeah especially as an immutable no need to go via a func call vs direct var access?